File: src/test/java/com/github/nsc/de/shake/interpreter/values/BooleanValueTest.java
package com.github.nsc.de.shake.interpreter.values;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.shake.interpreter.values.BooleanValue.*;

public class BooleanValueTest {

    @Test
    public void testFromBoolean() {

        assertSame(TRUE, from(true));
        assertSame(FALSE, from(false));

    }

    @Test
    public void testFromInterpreterValue() {

        // Boolean-values
        assertSame(TRUE, from(TRUE));
        assertSame(FALSE, from(FALSE));

        // Integers
        assertSame(TRUE, from(IntegerValue.ONE));
        assertSame(TRUE, from(new IntegerValue(Integer.MAX_VALUE)));
        assertSame(FALSE, from(new IntegerValue(Integer.MIN_VALUE)));
        assertSame(FALSE, from(new IntegerValue(0)));

        // Double
        assertSame(TRUE, from(new DoubleValue(1)));
        assertSame(TRUE, from(new DoubleValue(Double.MAX_VALUE)));
        assertSame(TRUE, from(new DoubleValue(Double.MIN_VALUE)));
        assertSame(TRUE, from(new DoubleValue(0.1)));
        assertSame(FALSE, from(new DoubleValue(-0.1)));
        assertSame(FALSE, from(new IntegerValue(0)));

        // Null
        assertSame(FALSE, from(NullValue.NULL));

        // TODO Function test

    }

    @Test
    public void testAdd() {
        Error error = assertThrows(Error.class, () -> TRUE.add(TRUE));
        assertEquals("Operator '+' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testSub() {
        Error error = assertThrows(Error.class, () -> TRUE.sub(TRUE));
        assertEquals("Operator '-' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testMul() {
        Error error = assertThrows(Error.class, () -> TRUE.mul(TRUE));
        assertEquals("Operator '*' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testDiv() {
        Error error = assertThrows(Error.class, () -> TRUE.div(TRUE));
        assertEquals("Operator '/' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testMod() {
        Error error = assertThrows(Error.class, () -> TRUE.mod(TRUE));
        assertEquals("Operator '%' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testPow() {
        Error error = assertThrows(Error.class, () -> TRUE.pow(TRUE));
        assertEquals("Operator '**' is not defined for type boolean", error.getMessage());
    }



    @Test
    public void testOr() {
        assertEquals(TRUE, TRUE.or(TRUE));
        assertEquals(TRUE, FALSE.or(TRUE));
        assertEquals(TRUE, TRUE.or(FALSE));
        assertEquals(FALSE, FALSE.or(FALSE));
    }

    @Test
    public void testAnd() {
        assertEquals(TRUE, TRUE.and(TRUE));
        assertEquals(FALSE, FALSE.and(TRUE));
        assertEquals(FALSE, TRUE.and(FALSE));
        assertEquals(FALSE, FALSE.and(FALSE));
    }



    @Test
    public void testEqualsEquals() {
        assertEquals(TRUE, TRUE.equals(TRUE));
        assertEquals(FALSE, FALSE.equals(TRUE));
        assertEquals(FALSE, TRUE.equals(FALSE));
        assertEquals(TRUE, FALSE.equals(FALSE));
    }

    @Test
    public void testBiggerEquals() {
        Error error = assertThrows(Error.class, () -> TRUE.bigger_equals(TRUE));
        assertEquals("Operator '>=' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testSmallerEquals() {
        Error error = assertThrows(Error.class, () -> TRUE.smaller_equals(TRUE));
        assertEquals("Operator '<=' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testBigger() {
        Error error = assertThrows(Error.class, () -> TRUE.bigger(TRUE));
        assertEquals("Operator '>' is not defined for type boolean", error.getMessage());
    }

    @Test
    public void testSmaller() {
        Error error = assertThrows(Error.class, () -> TRUE.smaller(TRUE));
        assertEquals("Operator '<' is not defined for type boolean", error.getMessage());
    }



}

File: src/test/kotlin/com/github/nsc/de/shake/interpreter/values/BooleanValueTest.kt
package com.github.nsc.de.shake.interpreter.values

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class BooleanValueTest {
    @Test
    fun testFromBoolean() {
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(true))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(false))
    }

    @Test
    fun testFromInterpreterValue() {

        // Boolean-values
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(BooleanValue.TRUE))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(BooleanValue.FALSE))

        // Integers
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(IntegerValue.ONE))
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(IntegerValue(Int.MAX_VALUE)))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(IntegerValue(Int.MIN_VALUE)))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(IntegerValue(0)))

        // Double
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(DoubleValue(1.0)))
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(DoubleValue(Double.MAX_VALUE)))
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(DoubleValue(Double.MIN_VALUE)))
        Assertions.assertSame(BooleanValue.TRUE, BooleanValue.from(DoubleValue(0.1)))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(DoubleValue(-0.1)))
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(IntegerValue(0)))

        // Null
        Assertions.assertSame(BooleanValue.FALSE, BooleanValue.from(NullValue.NULL))

        // TODO Function test
    }

    @Test
    fun testAdd() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.add(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '+' is not defined for type boolean", error.message)
    }

    @Test
    fun testSub() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.sub(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '-' is not defined for type boolean", error.message)
    }

    @Test
    fun testMul() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.mul(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '*' is not defined for type boolean", error.message)
    }

    @Test
    fun testDiv() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.div(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '/' is not defined for type boolean", error.message)
    }

    @Test
    fun testMod() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.mod(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '%' is not defined for type boolean", error.message)
    }

    @Test
    fun testPow() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.pow(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '**' is not defined for type boolean", error.message)
    }

    @Test
    fun testOr() {
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.TRUE.or(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.FALSE.or(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.TRUE.or(BooleanValue.FALSE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.FALSE.or(BooleanValue.FALSE))
    }

    @Test
    fun testAnd() {
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.TRUE.and(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.FALSE.and(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.TRUE.and(BooleanValue.FALSE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.FALSE.and(BooleanValue.FALSE))
    }

    @Test
    fun testEqualsEquals() {
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.TRUE.equals(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.FALSE.equals(BooleanValue.TRUE))
        Assertions.assertEquals(BooleanValue.FALSE, BooleanValue.TRUE.equals(BooleanValue.FALSE))
        Assertions.assertEquals(BooleanValue.TRUE, BooleanValue.FALSE.equals(BooleanValue.FALSE))
    }

    @Test
    fun testBiggerEquals() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.bigger_equals(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '>=' is not defined for type boolean", error.message)
    }

    @Test
    fun testSmallerEquals() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.smaller_equals(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '<=' is not defined for type boolean", error.message)
    }

    @Test
    fun testBigger() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.bigger(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '>' is not defined for type boolean", error.message)
    }

    @Test
    fun testSmaller() {
        val error = Assertions.assertThrows(Error::class.java) { BooleanValue.TRUE.smaller(BooleanValue.TRUE) }
        Assertions.assertEquals("Operator '<' is not defined for type boolean", error.message)
    }
}